drivers/dma/dma_cavs.c

```diff
@@ -0,0 +1,454 @@
+#include <board.h>
+#include <device.h>
+#include <init.h>
+#include <dma.h>
+
+#include "dma_cavs.h"
+
+#define SYS_LOG_DOMAIN "dev/dma_cavs"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_DMA_LEVEL
+#include <logging/sys_log.h>
+
+#define BYTE				(1)
+#define WORD				(2)
+#define DWORD				(4)
+
+/* CFG_LO */
+#define DW_CFG_CLASS(x)			(x << 29)
+/* CFG_HI */
+#define DW_CFGH_SRC_PER(x)		((x & 0xf) | ((x & 0x30) << 24))
+#define DW_CFGH_DST_PER(x)		(((x & 0xf) << 4) | ((x & 0x30) << 26))
+
+/* default initial setup register values */
+#define DW_CFG_LOW_DEF			0x0
+
+#define DEV_NAME(dev) ((dev)->config->name)
+#define DEV_DATA(dev) ((struct dw_dma_dev_data *const)(dev)->driver_data)
+#define DEV_CFG(dev) \
+	((const struct dw_dma_dev_cfg *const)(dev)->config->config_info)
+
+/* number of tries to wait for reset */
+#define DW_DMA_CFG_TRIES	10000
+#define INT_MASK_ALL		0xFF00
+
+static ALWAYS_INLINE void dw_write(u32_t dma_base, u32_t reg, u32_t value)
+{
+	*((volatile u32_t*)(dma_base + reg)) = value;
+}
+
+static ALWAYS_INLINE u32_t dw_read(u32_t dma_base, u32_t reg)
+{
+	return *((volatile u32_t*)(dma_base + reg));
+}
+
+static void dw_dma_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct dw_dma_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	struct dw_dma_dev_data *const dev_data = DEV_DATA(dev);
+	struct dma_chan_data *chan_data;
+
+	u32_t status_tfr = 0;
+	u32_t status_block = 0;
+	u32_t status_err = 0;
+	u32_t status_intr;
+	u32_t channel;
+
+	status_intr = dw_read(dev_cfg->base, DW_INTR_STATUS);
+	if (!status_intr) {
+		SYS_LOG_ERR("status_intr = %d", status_intr);
+	}
+
+	/* get the source of our IRQ. */
+	status_block = dw_read(dev_cfg->base, DW_STATUS_BLOCK);
+	status_tfr = dw_read(dev_cfg->base, DW_STATUS_TFR);
+
+	/* TODO: handle errors, just clear them atm */
+	status_err = dw_read(dev_cfg->base, DW_STATUS_ERR);
+	if (status_err) {
+		SYS_LOG_ERR("status_err = %d\n", status_err);
+		dw_write(dev_cfg->base, DW_CLEAR_ERR, status_err);
+	}
+
+	/* clear interrupts */
+	dw_write(dev_cfg->base, DW_CLEAR_BLOCK, status_block);
+	dw_write(dev_cfg->base, DW_CLEAR_TFR, status_tfr);
+
+	/* Dispatch ISRs for channels depending upon the bit set */
+	if (dev_data->dma_blkcallback) {
+		while (status_block) {
+			channel = find_lsb_set(status_block) - 1;
+			status_block &= ~(1 << channel);
+			chan_data = &dev_data->dw_pdata.chan[channel];
+			k_free(chan_data->lli);
+			dev_data->dma_blkcallback(dev, channel, 0);
+		}
+	}
+
+	if (dev_data->dma_tfrcallback) {
+		while (status_tfr) {
+			channel = find_lsb_set(status_tfr) - 1;
+			status_tfr &= ~(1 << channel);
+			chan_data = &dev_data->dw_pdata.chan[channel];
+			k_free(chan_data->lli);
+			dev_data->dma_tfrcallback(dev, channel, 0);
+		}
+	}
+}
+
+static int dw_dma_config(struct device *dev, u32_t channel,
+			 struct dma_config *cfg)
+{
+	struct dw_dma_dev_data *const dev_data = DEV_DATA(dev);
+	struct dma_chan_data *chan_data = &dev_data->dw_pdata.chan[channel];
```

Comment:
Please check if channel is within bounds before taking a pointer to the chan array.

Reply:
ok.